src/jltypes.c

```diff
@@ -1687,17 +1687,43 @@ jl_value_t *jl_apply_type_(jl_value_t *tc, jl_value_t **params, size_t n)
         }
     }
     if (tc == (jl_value_t*)jl_ntuple_type && (n == 1 || n == 2)) {
-        if (jl_is_long(params[0])) {
-            size_t nt = jl_unbox_long(params[0]);
+        if (!jl_is_typevar(params[0])) {
+            size_t nt;
+            if (jl_is_long(params[0])) {
+                nt = jl_unbox_long(params[0]);
+            }
+#ifdef _P64
+            else if (jl_is_uint64(params[0])) {
+                nt = jl_unbox_uint64(params[0]);
+            }
+            else if (jl_is_int32(params[0])) {
+                nt = jl_unbox_int32(params[0]);
+            }
+#endif
+            else if (jl_is_uint32(params[0])) {
+                nt = jl_unbox_uint32(params[0]);
+            }
+            else if (jl_is_uint16(params[0])) {
+                nt = jl_unbox_uint16(params[0]);
+            }
+            else if (jl_is_uint8(params[0])) {
+                nt = jl_unbox_uint8(params[0]);
+            }
+            else if (jl_is_int16(params[0])) {
+                nt = jl_unbox_int16(params[0]);
+            }
+            else if (jl_is_int8(params[0])) {
+                nt = jl_unbox_int8(params[0]);
+            }
+            else if (!jl_is_typevar(params[0])) {
+                // Only allow integer, unsigned or TypeVar as the first parameter to
+                // NTuple. issue #9233
+                jl_type_error_rt("apply_type", "first parameter of NTuple",
+                                 (jl_value_t*)jl_long_type, params[0]);
```

Comment:
What's the wrong parameter? The only thing I didn't see how do to, is to have the error say "Integer" instead of "Int64" (or "Int32" on a 32-bit machine).